File: src/julia/explainers/es.jl
import ESValues

type ESExplainer <: Explainer
    model::Model
    data::Data
    link::Link
    nsamples::Int64
    explain::Function
end
ESExplainer(model, data) = ESExplainer(model, data, IdentityLink())
ESExplainer(model, data, link) = ESExplainer(model, data, link, 0)
function ESExplainer(model, data, link, nsamples)
    e = ESExplainer(convert(Model, model), convert(Data, data), convert(Link, link), nsamples, x->x)
    match_data!(e.model, e.data)
    e.explain = x->explain(convert(Instance, x), e)
    e
end

function explain(instance::Instance, e::ESExplainer)
    local fx

    # make sure the instance is compatible with the background data
    match_data!(instance, e.data)

    # make sure the functions we got from the user actually work
    try
        e.model.f(e.data.data)
    catch err
        error("Provided model function fails when applied to the provided data set: ", err)
    end
    try
        fx = e.model.f(instance.x)
        length(fx) == 1 || throw(Exception("Length of model.f(x) should be 1"))
    catch err
        error("Provided model function fails when applied to the provided data instance x: ", err)
    end

    baseValue,effects,effectsVar = ESValues.esvalues(
        e.data.transposed ? instance.x : instance.x',
        e.data.transposed ? e.model.f : x->e.model.f(x'),
        e.data.transposed ? e.data.data : e.data.data',
        e.link.f;
        featureGroups=e.data.groups,
        weights=e.data.weights,
        nsamples=e.nsamples
    )
    AdditiveExplanation(e.link.f(baseValue), fx, effects, effectsVar, instance, e.link, e.model, e.data)
end

# function explain(model::Model, data::Data, e::ESExplainer)
#     if data.transposed
#         return Explanation[explain(data.data[:,[i]], model::Model, data::Data, e::ESExplainer) for i in 1:size(data.data)[2]]
#     else
#         return Explanation[explain(data.data[[i],:], model::Model, data::Data, e::ESExplainer) for i in 1:size(data.data)[1]]
#     end
# end

File: src/python/explainers/es.py
import esvalues
from ..common import convert_to_instance, convert_to_model, match_instance_to_data, match_model_to_data
from ..explanations import AdditiveExplanation
from ..links import convert_to_link, IdentityLink
from ..datatypes import convert_to_data


class ESExplainer:
    def __init__(self, model, data, link=IdentityLink(), **kwargs):
        self.link = convert_to_link(link)
        self.model = convert_to_model(model)
        self.data = convert_to_data(data)
        match_model_to_data(self.model, self.data)
        self.estimator = esvalues.ESValuesEstimator(self.model.f, self.data.data, self.link.f, **kwargs)

    def explain(self, instance):
        iobj = convert_to_instance(instance)
        match_instance_to_data(iobj, self.data)
        fnull, phi, phiVar = self.estimator.esvalues(iobj.x)
        return AdditiveExplanation(fnull, phi, phiVar, iobj, self.link, self.model, self.data)
